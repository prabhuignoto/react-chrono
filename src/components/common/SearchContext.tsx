import React, { createContext, useState, useContext } from 'react';

interface SearchContextType {
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

// Create a context with default values
const SearchContext = createContext<SearchContextType>({
  searchTerm: '',
  setSearchTerm: () => {},
});

/**
 * Provider component for the search context
 */
export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <SearchContext.Provider value={{ searchTerm, setSearchTerm }}>
      {children}
    </SearchContext.Provider>
  );
};

/**
 * Custom hook to use the search context
 */
export const useSearch = (): SearchContextType => {
  const context = useContext(SearchContext);

  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }

  return context;
};

export default SearchContext;
