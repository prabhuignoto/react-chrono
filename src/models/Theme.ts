export interface Theme {
  cardBgColor?: string;
  cardForeColor?: string;
  primary?: string;
  secondary?: string;
  textColor?: string;
  titleColor?: string;
}
