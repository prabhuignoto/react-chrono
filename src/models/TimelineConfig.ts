/**
 * New grouped configuration interfaces for better API organization
 * This provides a cleaner, more intuitive way to configure the timeline
 */

import { ReactNode } from 'react';

/**
 * Layout and sizing configuration
 */
export interface LayoutConfig {
  /** Maximum width of timeline cards in pixels */
  cardWidth?: number;
  
  /** Minimum height of timeline cards in pixels */
  cardHeight?: number;
  
  /** Size of timeline points in pixels */
  pointSize?: number;
  
  /** Width of the timeline track line in pixels */
  lineWidth?: number;
  
  /** Width of each timeline section in horizontal mode */
  itemWidth?: number;
  
  /** Height of the timeline container (number = px or CSS length string) */
  timelineHeight?: number | string;
  
  /** Responsive behavior configuration */
  responsive?: {
    /** Viewport width breakpoint for switching modes */
    breakpoint?: number;
    /** Whether responsive breakpoint switching is enabled */
    enabled?: boolean;
  };
  
  /** Positioning and layout options */
  positioning?: {
    /** Card position in horizontal mode */
    cardPosition?: 'top' | 'bottom';
    /** Flip layout for RTL support */
    flipLayout?: boolean;
  };
}

/**
 * User interaction and navigation configuration
 */
export interface InteractionConfig {
  /** Enable keyboard navigation with arrow keys */
  keyboardNavigation?: boolean;
  
  /** Enable clicking on timeline points */
  pointClick?: boolean;
  
  /** Enable auto-scroll to active items */
  autoScroll?: boolean;
  
  /** Focus active item on component load */
  focusOnLoad?: boolean;
  
  /** Highlight cards on mouse hover */
  cardHover?: boolean;
  
  /** Disable all user interactions */
  disabled?: boolean;
}

/**
 * Content handling and display configuration
 */
export interface ContentConfig {
  /** Allow HTML parsing in card content */
  allowHTML?: boolean;
  
  /** Enable read more functionality for long content */
  readMore?: boolean;
  
  /** Display text as overlay on media */
  textOverlay?: boolean;
  
  /** Date format for timeline titles */
  dateFormat?: string;
  
  /** Use compact text display */
  compactText?: boolean;
  
  /** Semantic HTML tags for accessibility */
  semanticTags?: {
    title?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'div';
    subtitle?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'div';
  };
}

/**
 * Visual display and styling configuration
 */
export interface DisplayConfig {
  /** Remove borders and shadows from cards */
  borderless?: boolean;
  
  /** Hide timeline cards completely */
  cardsDisabled?: boolean;
  
  /** Hide timeline points */
  pointsDisabled?: boolean;
  
  /** Shape of timeline points */
  pointShape?: 'circle' | 'square' | 'diamond';
  
  /** Show all cards simultaneously in horizontal mode */
  allCardsVisible?: boolean;
  
  /** Toolbar configuration */
  toolbar?: {
    /** Enable/disable the toolbar */
    enabled?: boolean;
    /** Position of the toolbar */
    position?: 'top' | 'bottom';
    /** Make toolbar sticky during scroll */
    sticky?: boolean;
  };
  
  /** Scrolling configuration */
  scrollable?: boolean | { 
    /** Show scrollbar */
    scrollbar: boolean; 
  };
}

/**
 * Media content configuration
 */
export interface MediaConfig {
  /** Minimum height of media elements in pixels */
  height?: number;
  
  /** Media alignment within cards */
  align?: 'left' | 'center' | 'right';
  
  /** CSS object-fit property for images */
  fit?: 'cover' | 'contain' | 'fill' | 'none' | 'scale-down';
}

/**
 * Animation and slideshow configuration
 */
export interface AnimationConfig {
  /** Slideshow configuration */
  slideshow?: {
    /** Enable slideshow functionality */
    enabled?: boolean;
    /** Duration each slide is displayed in milliseconds */
    duration?: number;
    /** Type of slideshow transition animation */
    type?: 'reveal' | 'slide' | 'fade';
    /** Auto-start slideshow on load */
    autoStart?: boolean;
    /** Show progress indicator on individual cards */
    showProgress?: boolean;
    /** Show overall progress bar */
    showOverallProgress?: boolean;
  };
}

/**
 * Styling and theming configuration
 */
export interface StyleConfig {
  /** Custom CSS class names for elements */
  classNames?: {
    card?: string;
    cardMedia?: string;
    cardSubTitle?: string;
    cardText?: string;
    cardTitle?: string;
    controls?: string;
    title?: string;
    timelinePoint?: string;
    timelineTrack?: string;
  };
  
  /** Custom font sizes */
  fontSizes?: {
    cardSubtitle?: string;
    cardText?: string;
    cardTitle?: string;
    title?: string;
  };
}

/**
 * Accessibility and internationalization
 */
export interface AccessibilityConfig {
  /** Custom button text labels */
  buttonTexts?: {
    first?: string;
    last?: string;
    next?: string;
    previous?: string;
    play?: string;
    stop?: string;
  };
  
  /** Search functionality labels */
  search?: {
    placeholder?: string;
    ariaLabel?: string;
    clearLabel?: string;
  };
}