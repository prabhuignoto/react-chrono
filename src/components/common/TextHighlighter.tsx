import React from 'react';
import styled from 'styled-components';
import { Theme } from '@models/Theme';

interface HighlightedTextProps {
  text: string;
  searchTerm: string;
  theme: Theme;
}

// Styled component for highlighted text
const HighlightSpan = styled.span<{ theme: Theme }>`
  background-color: ${({ theme }) =>
    theme.primary}33; /* Primary color with 20% opacity */
  color: ${({ theme }) => theme.primary};
  font-weight: 500;
  border-radius: 2px;
  padding: 0 2px;
`;

/**
 * Component that highlights occurrences of a search term within text
 */
const TextHighlighter: React.FC<HighlightedTextProps> = ({
  text,
  searchTerm,
  theme,
}) => {
  // If no search term or no text, just return the original text
  if (!searchTerm || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the search term
  const escapedSearchTerm = searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Create a case-insensitive regular expression to find all occurrences
  const regex = new RegExp(`(${escapedSearchTerm})`, 'gi');

  // Split the text by the search term
  const parts = text.split(regex);

  // Return the text with highlighted parts
  return (
    <>
      {parts.map((part, i) =>
        regex.test(part) ? (
          <HighlightSpan key={i} theme={theme}>
            {part}
          </HighlightSpan>
        ) : (
          <React.Fragment key={i}>{part}</React.Fragment>
        ),
      )}
    </>
  );
};

export default TextHighlighter;
