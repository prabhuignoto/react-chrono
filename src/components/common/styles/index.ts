import { Theme } from '@models/Theme';
import { css } from 'styled-components';

export const ScrollBar = css<{ theme: Theme }>`
  scrollbar-color: ${(p) => p.theme?.primary} default;
  scrollbar-width: thin;

  &::-webkit-scrollbar {
    width: 0.3em;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.2);
  }

  &::-webkit-scrollbar-thumb {
    background-color: ${(p) => p.theme?.primary};
    outline: 1px solid ${(p) => p.theme?.primary};
  }
`;
