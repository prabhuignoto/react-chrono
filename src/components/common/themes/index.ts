import { Theme } from '@models/Theme';

export const defaultTheme: Theme = {
  cardBgColor: '#ffffff',
  cardDetailsBackGround: '#ffffff',
  cardDetailsColor: '#000',
  cardMediaBgColor: '#f5f5f5',
  cardSubtitleColor: '#000',
  cardTitleColor: '#007FFF',
  detailsColor: '#000',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#f5f5f5',
  nestedCardDetailsBackGround: '#f5f5f5',
  nestedCardDetailsColor: '#000',
  nestedCardSubtitleColor: '#000',
  nestedCardTitleColor: '#000',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#f5f5f5',
  toolbarBtnBgColor: '#fff',
  toolbarTextColor: '#000',
};

export const darkTheme: Theme = {
  cardBgColor: '#191919',
  cardDetailsBackGround: '#191919',
  cardDetailsColor: '#ffff0f',
  cardMediaBgColor: '#2f2f2f',
  cardSubtitleColor: '#ffffff',
  cardTitleColor: '#007FFF',
  detailsColor: '#ffffff',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#333333',
  nestedCardDetailsBackGround: '#333333',
  nestedCardDetailsColor: '#ffffff',
  nestedCardSubtitleColor: '#ffffff',
  nestedCardTitleColor: '#ffffff',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#555',
  toolbarBtnBgColor: '#222',
  toolbarTextColor: '#ffffff',
};
