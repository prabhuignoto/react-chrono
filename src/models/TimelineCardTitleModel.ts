import { Theme } from './Theme';

export interface TitleModel {
  active?: boolean;
  align?: 'left' | 'right';
  classString?: string;
  theme?: Theme;
  title?: string;
}
