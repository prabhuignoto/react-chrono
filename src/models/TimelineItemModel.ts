import { ReactNode } from 'react';
import { Theme } from './Theme';
import { Scroll } from './TimelineHorizontalModel';
import { Media } from './TimelineMediaModel';
import { TimelineProps } from './TimelineModel';

/**
 * Represents the model for a timeline item.
 */
export interface TimelineItemModel {
  // Internal property for handling date.
  _dayjs?: any;

  // Indicates if the timeline item is active.
  active?: boolean;

  // Detailed text for the timeline card.
  cardDetailedText?: string | string[];

  // Subtitle for the timeline card.
  cardSubtitle?: string;

  // Title for the timeline card.
  cardTitle?: string;

  // Main content of the timeline item.
  content?: ReactNode | ReactNode[];

  // Date associated with the timeline item.
  date?: number | string | Date;

  // Unique identifier for the timeline item.
  id?: string;

  // Indicates if the timeline item is nested.
  isNested?: boolean;

  // Array of nested timeline items.
  items?: TimelineItemModel[];

  // Media associated with the timeline item.
  media?: Media;

  // Position of the timeline item.
  position?: string;

  // Custom content for the timeline content.
  timelineContent?: ReactNode;

  // Title of the timeline item.
  title?: string;

  // URL associated with the timeline item.
  url?: string;

  // Indicates if the timeline item is visible.
  visible?: boolean;
}

/**
 * Represents the model for a timeline card.
 */
export type TimelineCardModel = Pick<
  TimelineItemModel,
  | 'id'
  | 'visible'
  | 'title'
  | 'active'
  | 'cardDetailedText'
  | 'cardSubtitle'
  | 'cardTitle'
  | 'media'
  | 'url'
  | 'timelineContent'
  | 'isNested'
  | 'items'
> & {
  // Function for auto-scrolling.
  autoScroll?: ({
    pointOffset,
    pointWidth,
    timelinePointHeight,
    contentHeight,
  }: Partial<Scroll>) => void;

  // Custom content for the timeline card.
  customContent?: React.ReactNode | React.ReactNode[];

  // Indicates if the timeline card has focus.
  hasFocus?: boolean;

  // Icon element for the timeline card.
  iconChild?: React.ReactNode;

  // Click event handler for the timeline card.
  onClick?: (id?: string) => void;

  // Elapsed event handler for the timeline card.
  onElapsed?: (id?: string) => void;

  // Duration of slide item transitions.
  slideItemDuration?: number;

  // Indicates if the slide show is running.
  slideShowRunning?: boolean;

  // Theme to be applied to the timeline card.
  theme?: Theme;

  // Unique identifier for the timeline wrapper.
  wrapperId: string;
} & Pick<TimelineProps, 'cardHeight' | 'cardWidth' | 'nestedCardHeight'>;
