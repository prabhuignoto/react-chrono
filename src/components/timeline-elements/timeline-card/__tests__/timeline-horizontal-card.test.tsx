import { afterEach, beforeEach, describe, vi } from 'vitest';
import { TimelineCardModel } from '../../../../models/TimelineItemModel';
import { customRender } from '../../../common/test';
import { providerProps } from '../../../common/test/index';
import TimelineCard from '../timeline-horizontal-card';

const defaultProps: TimelineCardModel = {
  active: true,
  cardDetailedText: 'card detailed text',
  cardSubtitle: 'card subtitle',
  cardTitle: 'card title',
  iconChild: 'icon child',
  id: 'id-12345',
  media: {
    name: 'image_new',
    source: {
      type: 'IMAGE',
      url: 'https://picsum.photos/200/300',
    },
    type: 'IMAGE',
  },
  onClick: vi.fn(),
  onElapsed: vi.fn(),
  wrapperId: 'wrapper-id-12345',
};

describe('TimelineHorizontalCard', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: vi.fn().mockImplementation((query) => ({
        addEventListener: vi.fn(),
        dispatchEvent: vi.fn(),
        matches: false,
        media: query,
        onchange: null,
        removeEventListener: vi.fn(),
      })),
      writable: true,
    });
  });

  beforeEach(() => {
    // create a div with id 'wrapper-id-12345' to render the component
    const root = document.createElement('div');
    root.setAttribute('id', 'wrapper-id-12345');
    document.body.appendChild(root);
  });

  afterEach(() => {
    // remove the div with id 'wrapper-id-12345' after the tests
    const root = document.getElementById('wrapper-id-12345');
    if (root) {
      document.body.removeChild(root);
    }
  });

  // should render timeline card
  it('should render timeline card', () => {
    const { container } = customRender(<TimelineCard {...defaultProps} />, {
      providerProps: { ...providerProps, cardLess: false },
    });

    expect(container).toBeInTheDocument();
  });

  // should match the snapshot
  it('should match the snapshot', () => {
    const { container } = customRender(<TimelineCard {...defaultProps} />, {
      providerProps: { ...providerProps, cardLess: false },
    });

    expect(container).toMatchSnapshot();
  });

  // should render cardDetailedText, cardSubtitle, cardTitle, iconChild
  it('should render cardDetailedText, cardSubtitle, cardTitle', () => {
    const { getByText } = customRender(<TimelineCard {...defaultProps} />, {
      providerProps,
    });

    expect(getByText('card detailed text')).toBeInTheDocument();
    expect(getByText('card subtitle')).toBeInTheDocument();
    expect(getByText('card title')).toBeInTheDocument();
  });

  // should support rendering a custom icon
  it('should render a custom icon', () => {
    const { getByText } = customRender(
      <TimelineCard
        {...defaultProps}
        iconChild={<span role="img">icon</span>}
      ></TimelineCard>,
      {
        providerProps,
      },
    );

    expect(getByText('icon')).toBeInTheDocument();
  });

  // should render media (image) if provided
  it('should render media (image) if provided', () => {
    const { getByAltText } = customRender(<TimelineCard {...defaultProps} />, {
      providerProps,
    });

    expect(getByAltText('image_new')).toBeInTheDocument();
  });

  // should render media (video) if provided
  it('should render media (video) if provided', () => {
    const { getByText, getByRole } = customRender(
      <TimelineCard {...defaultProps} url="http://www.google.com" />,
      {
        providerProps,
      },
    );

    expect(getByText('card title')).toBeInTheDocument();
    expect(getByText('card subtitle')).toBeInTheDocument();
    expect(getByRole('link')).toBeInTheDocument();
    expect(getByRole('link')).toHaveAttribute('href', 'http://www.google.com');
  });
});
