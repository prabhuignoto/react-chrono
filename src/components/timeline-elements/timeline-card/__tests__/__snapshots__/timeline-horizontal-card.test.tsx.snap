// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimelineHorizontalCard > should match the snapshot 1`] = `
<div>
  <div
    class="sc-hAtEyd bvLluf vertical_alternating"
    data-testid="timeline-item"
  >
    <div
      class="sc-fGFwAa jXmrA-D"
    >
      <div
        class="sc-blLsxD hlnyEA timeline-circle using-icon vertical_alternating active"
        data-testid="timeline-circle"
      >
        icon child
      </div>
    </div>
    <div
      class="sc-ieZDjg dhUWhc vertical_alternating"
      data-testid="timeline-title"
    >
      <div
        class="sc-ktEKTO kUkWyA timeline-item-title active title"
        font-size="1.5rem"
      />
    </div>
  </div>
</div>
`;
